File: scripts/sandbox_wrapper.py
#!/usr/bin/env python3
"""
Genoma Sandbox Wrapper — Python
Executes a user script in the sandbox protocol format.
Communication: stdin (JSON) → script execution → stdout (JSON-lines)
"""
import sys
import json
import traceback
import importlib.util
import io
from contextlib import redirect_stdout, redirect_stderr


def emit(msg_type: str, data, tb: str = ""):
    """Write a protocol message to stdout."""
    msg = {"type": msg_type, "data": data}
    if tb:
        msg["traceback"] = tb
    print(json.dumps(msg), flush=True)


def main():
    try:
        # Read input from stdin
        raw_input = sys.stdin.read()
        if not raw_input.strip():
            emit("error", "No input received on stdin")
            sys.exit(1)

        payload = json.loads(raw_input)
        script_path = payload.get("script_path", "/workspace/script.py")
        user_input = payload.get("input", {})

        # Load the user script as a module
        spec = importlib.util.spec_from_file_location("user_script", script_path)
        if spec is None:
            emit("error", f"Cannot load script: {script_path}")
            sys.exit(1)

        module = importlib.util.module_from_spec(spec)

        # Capture stdout/stderr from user script
        captured_stdout = io.StringIO()
        captured_stderr = io.StringIO()

        # Inject input as a global variable
        module.__dict__["INPUT"] = user_input
        module.__dict__["emit_log"] = lambda msg: emit("log", str(msg))

        with redirect_stdout(captured_stdout), redirect_stderr(captured_stderr):
            spec.loader.exec_module(module)

        # Log captured output
        stdout_val = captured_stdout.getvalue()
        if stdout_val.strip():
            for line in stdout_val.strip().split("\n"):
                emit("log", line)

        stderr_val = captured_stderr.getvalue()
        if stderr_val.strip():
            for line in stderr_val.strip().split("\n"):
                emit("log", f"[stderr] {line}")

        # Get result from the module
        if hasattr(module, "RESULT"):
            result = module.RESULT
        elif hasattr(module, "main"):
            result = module.main(user_input)
        else:
            result = {"status": "completed"}

        # Ensure result is serializable
        if not isinstance(result, dict):
            result = {"value": result}

        emit("result", result)

    except json.JSONDecodeError as e:
        emit("error", f"Invalid JSON input: {e}")
        sys.exit(1)
    except Exception as e:
        emit("error", str(e), traceback.format_exc())
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/sandbox_wrapper.js
#!/usr/bin/env node
/**
 * Genoma Sandbox Wrapper — Node.js
 * Executes a user script in the sandbox protocol format.
 * Communication: stdin (JSON) → script execution → stdout (JSON-lines)
 */

const fs = require('fs');
const path = require('path');
const { Module } = require('module');

function emit(type, data, traceback = '') {
    const msg = { type, data };
    if (traceback) msg.traceback = traceback;
    process.stdout.write(JSON.stringify(msg) + '\n');
}

async function main() {
    try {
        // Read input from stdin
        const chunks = [];
        for await (const chunk of process.stdin) {
            chunks.push(chunk);
        }
        const rawInput = Buffer.concat(chunks).toString('utf-8').trim();

        if (!rawInput) {
            emit('error', 'No input received on stdin');
            process.exit(1);
        }

        const payload = JSON.parse(rawInput);
        const scriptPath = payload.script_path || '/workspace/script.js';
        const userInput = payload.input || {};

        // Verify script exists
        if (!fs.existsSync(scriptPath)) {
            emit('error', `Script not found: ${scriptPath}`);
            process.exit(1);
        }

        // Capture console.log output
        const logs = [];
        const originalLog = console.log;
        const originalError = console.error;
        const originalWarn = console.warn;

        console.log = (...args) => {
            const msg = args.map(a => typeof a === 'string' ? a : JSON.stringify(a)).join(' ');
            emit('log', msg);
        };
        console.error = (...args) => {
            const msg = args.map(a => typeof a === 'string' ? a : JSON.stringify(a)).join(' ');
            emit('log', `[stderr] ${msg}`);
        };
        console.warn = console.error;

        // Make INPUT available globally
        global.INPUT = userInput;
        global.emit_log = (msg) => emit('log', String(msg));

        // Load and execute the script
        const scriptContent = fs.readFileSync(scriptPath, 'utf-8');

        // Create a module sandbox
        const scriptModule = new Module(scriptPath);
        scriptModule.filename = scriptPath;
        scriptModule.paths = Module._nodeModulePaths(path.dirname(scriptPath));

        // Execute
        const wrapper = Module.wrap(scriptContent);
        const compiled = require('vm').runInThisContext(wrapper, { filename: scriptPath });
        compiled.call(scriptModule.exports, scriptModule.exports, require, scriptModule, scriptPath, path.dirname(scriptPath));

        // Get result
        let result;
        if (typeof scriptModule.exports === 'function') {
            result = await scriptModule.exports(userInput);
        } else if (scriptModule.exports && typeof scriptModule.exports.main === 'function') {
            result = await scriptModule.exports.main(userInput);
        } else if (scriptModule.exports && typeof scriptModule.exports !== 'object') {
            result = { value: scriptModule.exports };
        } else if (scriptModule.exports.RESULT !== undefined) {
            result = scriptModule.exports.RESULT;
        } else {
            result = { status: 'completed' };
        }

        if (typeof result !== 'object' || result === null) {
            result = { value: result };
        }

        emit('result', result);

        // Restore console
        console.log = originalLog;
        console.error = originalError;
        console.warn = originalWarn;

    } catch (err) {
        emit('error', err.message, err.stack);
        process.exit(1);
    }
}

main();
